db/compaction/compaction.cc

```diff
@@ -415,6 +415,52 @@ void Compaction::MarkFilesBeingCompacted(bool mark_as_compacted) {
   }
 }
 
+Slice Compaction::StartLevelSmallestInternalKey() const {
+  // lazy initialize, return if it's already calculated
+  if (!start_level_smallest_internal_.empty()) {
+    return start_level_smallest_internal_;
+  }
+
+  assert(inputs_[0].level == start_level());
+  assert(!inputs_[0].files.empty());
+  // non-L0 inputs are sorted
+  start_level_smallest_internal_ = inputs_[0].files.front()->smallest.Encode();
+  if (start_level() == 0) {
+    const InternalKeyComparator icmp =
+        column_family_data()->internal_comparator();
+    for (const auto& file : inputs_[0].files) {
```

Comment:
The range might be a little bit too range so there is a problem: a large amount of KVs that have to be kept in hot tiers would be moved to the cold tier. For example, consider such a full compaction case:
```
L1: [key_3, key_5]
L2: [key_2, key_9]
L3: [......]
```
keys in L2 could all be determined to be cold and they should always be allowed to stay in L2. however, the range constraint of L1 would force them to be moved to L3. Not sure whether my understanding is correct.

Reply:
This is only for leveled compaction, my understanding is the inputs can only come from 2 levels, right? Or is this a valid level compaction: `L1 + L2 => L3`? (maybe we can handle the leveled compaction with data move up later, I'll add TODO for that).

Reply:
In that case, do we need to deal with L0? Can we skip L0?

Reply:
We changed to only use the feature for the last level compaction. And the safe range for moving up is changed.